### PR TITLE
Refactor jtx::Env

### PR DESCRIPTION
**Pull request by vinniefalco**
Some changes to support mao

Reviewers: @ximinez @scottschurr

### CODE REVIEW

**Comment by sublimator** [comment]
97% coverage. Legit!

**Comment by vinniefalco** [comment]
@sublimator  LOL...well...its not really fair because `jtx::Env` is used in all unit tests so of course its going to have almost full coverage!

File: src/ripple/test/impl/ManualTimeKeeper.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    time_point
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    return now_ + closeOffset_;
+}
+
+void
+ManualTimeKeeper::adjustCloseTime (duration amount)
+{
+    // Copied from TimeKeeper::adjustCloseTime
```
Comment (ximinez) [inline src/ripple/test/impl/ManualTimeKeeper.cpp]: To be pedantic, this is copied from `TimeKeeperImp::adjustCloseTime`.

File: src/ripple/test/impl/ManualTimeKeeper.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+auto
+ManualTimeKeeper::adjust(
+        std::chrono::system_clock::time_point when) ->
+    time_point
+{
+    return time_point(
+        std::chrono::duration_cast<duration>(
+            when.time_since_epoch() -
+                days(10957)));
```
Comment (ximinez) [inline src/ripple/test/impl/ManualTimeKeeper.cpp]: Rule of 3: We now how 3 instances of this constant (plus one in a comment). Should it be made into a global (`static`) variable? I would think it fits nicely in `NetClock`.

File: src/ripple/test/jtx/impl/Env.cpp [hunk 3]
```diff
@@ -60,8 +60,11 @@ Env::AppBundle::AppBundle(Application* app_)
 Env::AppBundle::AppBundle(std::unique_ptr<Config const> config)
 {
     auto logs = std::make_unique<Logs>();
-    owned = make_Application(
-        std::move(config), std::move(logs));
+    auto timeKeeper_ =
+        std::make_unique<ManualTimeKeeper>();
+    timeKeeper = timeKeeper_.get();
```
Comment (ximinez) [inline src/ripple/test/jtx/impl/Env.cpp]: Note to other reviewers: `timeKeeper` isn't referenced anywhere but here now, but there are future plans for it. eg. https://github.com/vinniefalco/rippled/commit/75532cb449716536c129e12e9cf2ad843be48837#diff-1fa292ad88b698cfe1b304837d40f7f9R124

Comment (ximinez) [inline src/ripple/test/jtx/impl/Env.cpp]: @vinniefalco So we have to be careful not to let `owned` get destroyed without clearing or updating `timeKeeper`, yes?

**Comment by ximinez** [comment]
Good setup. :+1:

**Comment by scottschurr** [comment]
I'm sorry to see `const` removed from all of the methods in `Env`.  It's not the end of the world, but I don't think it's the right choice.  I'll attempt to describe why.

Consider this statement from the isocpp FAQ (https://isocpp.org/wiki/faq/const-correctness#const-member-fns):

> What is a "`const` member function"?
> A member function that inspects (rather than mutates) its object.

In the following paragraphs there is no mention of whether the object holding the member function is `const`.  That's not relevant.  What matters is the promise that particular method makes.  So even though no one may ever make a `const Env`, there is still value in identifying a member function that will
not modify the logical state of that `Env` instance.

I just thought I should make my position clear.  Do whatever you think is correct.

**Comment by scottschurr** [comment]
The code looks swell.  There's clearly some kind of a setup going on with `TimeKeeper` that I don't understand yet.  But I trust that you and Ed have that figured out.  Looks good to me.
:+1:

**Comment by vinniefalco** [comment]
@scottschurr As more members of `Env` go through the `Application` object to achieve their implementation, it becomes less and less likely that even a seemingly read-only accessor function can be marked `const`. For example, many routines in rippled will update the "load manager" or "resource manager" even on observers. And these routines are certainly no longer thread safe (one of the promises C++11 offers for functions marked `const`). Given these facts, marking them `const` looks like a mistake.

**Comment by scottschurr** [comment]
@vinniefalco That's a fair explanation.  Thanks.

**Comment by vinniefalco** [comment]
Closing this because there are issues with the code that comes next

**Pull request closed without merging**